packages/opencensus-web-instrumentation-perf/src/initial-load-root-span.ts

```diff
@@ -0,0 +1,126 @@
+import {GroupedPerfEntries} from './perf-recorder';
+import {PerformanceLongTaskTiming, PerformanceNavigationTimingExtended} from './perf-types';
+import {getResourceSpan} from './resource-span';
+import {annotationsForPerfTimeFields} from './util';
+
+/**
+ * These are properties of PerformanceNavigationTiming that will be turned
+ * into span annotations on the navigation span.
+ */
+const NAVIGATION_TIMING_EVENTS = [
+  'domLoading',
+  'domInteractive',
+  'domContentLoaded',
+  'domComplete',
+  'loadEventStart',
+  'loadEventEnd',
+  'unloadEventStart',
+  'unloadEventEnd',
+];
+
+/**
+ * Returns a root span for the initial page load along with child spans for
+ * resource timings and long tasks that were part of the load. The root span
+ * represents the full time between when the navigation was initiated in the
+ * browser to when the `load` event fires.
+ * @param tracer The tracer to associate the root span with
+ * @param perfEntries Performance timing entries grouped by type. These should
+ *     include the entries up to soon after the `load` browser event fires.
+ * @param navigationFetchSpanId This is the ID for the span that represents the
+ *     HTTP request to get the initial HTML. This should be sent back from the
+ *     server to enable linking the server and client side spans for the initial
+ *     HTML fetch.
+ * @param traceId The trace ID that all spans returned should have. This should
+ *     also be specified by the server to enable linking between the server and
+ *     client spans for the initial HTML fetch.
+ */
+export function getInitialLoadRootSpan(
+    tracer: Tracer, perfEntries: GroupedPerfEntries,
+    navigationFetchSpanId: string, traceId: string): RootSpan {
+  const navTiming = perfEntries.navigationTiming;
+  const navigationUrl = navTiming ? navTiming.name : location.href;
+  const parsedNavigationUrl = parseUrl(navigationUrl);
+  const navigationPath = parsedNavigationUrl.pathname;
+  const root = new RootSpan(
+      tracer,
+      {name: `Nav.${navigationPath}`, spanContext: {spanId: '', traceId}});
+  root.startPerfTime = 0;
+  root.annotations = getNavigationAnnotations(perfEntries);
+  root.attributes[ATTRIBUTE_HTTP_URL] = navigationUrl;
+  root.attributes[ATTRIBUTE_HTTP_USER_AGENT] = navigator.userAgent;
+
+  if (navTiming) {
+    root.endPerfTime = navTiming.loadEventEnd;
+    root.attributes[ATTRIBUTE_NAV_TYPE] = navTiming.type;
+    const navFetchSpan =
+        getNavigationFetchSpan(navTiming, navigationUrl, navigationFetchSpanId);
+    root.spans.push(navFetchSpan);
+  }
+
+  const resourceSpans = perfEntries.resourceTimings.map(getResourceSpan);
+  const longTaskSpans = perfEntries.longTasks.map(getLongTaskSpan);
+
+  root.spans = root.spans.concat(resourceSpans, longTaskSpans);
+  for (const span of root.spans) {
+    span.traceId = traceId;
```

Comment:
Is it possible to pass ```traceId``` and ```root.id``` while constructing spans (```resourceSpans``` and ```longTaskSpans```), instead of looping here again?

Reply:
Sure, I added those parameters to the relevant functions and removed this loop.